helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterContext.java

```diff
@@ -104,6 +106,15 @@
     _estimatedMaxTopStateCount = estimateAvgReplicaCount(totalTopStateReplicas, instanceCount);
     _baselineAssignment = baselineAssignment;
     _bestPossibleAssignment = bestPossibleAssignment;
+
+    _busyInstances = _bestPossibleAssignment.values().stream().flatMap(
```

Comment:
Why use bestPossibleAssignment?  BestPossibleAssignment may not be the current cluster mapping. Shouldn't the `isBusy` boolean information determined during the initNewBatch in ClusterModelProvider?

Reply:
Current mapping is never directly used in the algorithm. To WAGED, the best possible assignment is treated as the current cluster mapping.

Given that saying, it cannot be determined by iniNewBatch. Since that input would already be processed by the rebalance scope calculating method. So it is not the one that we want.

Reply:
Is it because of yesterday's my question: if the cluster's present (current mapping) is a determined intermediate result from the past's (init mapping) calculation, the future of the cluster (final mapping) needs to be determined by the "past"'s calculation **best possible mapping** rather than current mapping? It sounds complex but I feel it's one of the trickiest things to understand and relate in code in the event-driven stateful system.

Reply:
If I get your question correctly, you are right. To be simple, the current state (which is intermediate) is never read in the whole rebalance calculation. It is mainly compared and processed by the other rebalance pipeline stages.